### PR TITLE
[JUJU-523] Ensure a robust isController wrapper for machine manifolds

### PR DESCRIPTION
**Pull request by manadart**
Some manifolds in the machine dependency engine are wrapped with `isController` to ensure they only run on controllers. This wrapper in turn has the state worker as a dependency.

Depending on state in this case is problematic, because if the state fails to ping Mongo, all such workers will bounce whether or not they depend on state.

Instead we use the state config watcher, which determines whether we are a controller based on the presence of state serving info. This will remain static when we change the Mongo primary (and bounce the state worker), leaving fewer workers to restart in such an event.

In addition, some redundant usage of `isController` is removed where it is already implied by other dependencies.

## QA steps

- Bootstrap and `juju enable-ha`.
- `juju deploy ubuntu`.
- Observe error free operation in the quiesced state.
- `juju ssh 0` and run `juju_engine_report`.
- Ensure that controller-only workers are not being run.
- `juju ssh -m controller 0` and run `juju_engine_report`.
- Ensure that controller-only workers are being run.
- Connect to the Mongo primary and run `rs.stepDown()`
- Observe the controller and workloads return to an error-free state.

In my simple A/B test, these workers did not bounce for a Mongo primary switch:
- agent-config-updater
- certificate-watcher
- fan-configurer (probably incidental)
- is-controller-flag
- model-cache-initialized-flag
- upgrade-database-flag
- upgrade-database-gate
- upgrade-database-runner

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

File: cmd/jujud/agent/machine/manifolds.go [hunk 1]
```diff
@@ -411,14 +411,14 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
 
 		// The multiwatcher manifold watches all the changes in the database
 		// through the AllWatcherBacking and manages notifying the multiwatchers.
-		multiwatcherName: ifDatabaseUpgradeComplete(ifController(multiwatcher.Manifold(multiwatcher.ManifoldConfig{
+		multiwatcherName: ifDatabaseUpgradeComplete(multiwatcher.Manifold(multiwatcher.ManifoldConfig{
```
Comment (jameinel) [inline cmd/jujud/agent/machine/manifolds.go:414]: Why would these be run on non-controllers?

File: cmd/jujud/agent/machine/manifolds.go [hunk 2]
```diff
@@ -432,15 +432,15 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
 		// The modelcache manifold creates a cache.Controller and keeps
 		// it up to date using an all model watcher. The controller is then
 		// used by the apiserver.
-		modelCacheName: ifDatabaseUpgradeComplete(ifController(modelcache.Manifold(modelcache.ManifoldConfig{
+		modelCacheName: ifDatabaseUpgradeComplete(modelcache.Manifold(modelcache.ManifoldConfig{
```
Comment (jameinel) [inline cmd/jujud/agent/machine/manifolds.go:435]: certainly I wouldn't expect all machine agents to be running the model cache

**Review by jameinel — COMMENTED** [review]
I don't understand the removal of isController but the changes for the flag looks good.

File: cmd/jujud/agent/machine/manifolds.go (hunk 2 above)
Comment (manadart) [reply cmd/jujud/agent/machine/manifolds.go:435]: `ifDatabaseUpgradeComplete` already implies this.

File: cmd/jujud/agent/machine/manifolds.go (hunk 1 above)
Comment (manadart) [reply cmd/jujud/agent/machine/manifolds.go:414]:
```
  multiwatcher:
    error: '"upgrade-database-flag" not running: dependency not available'
    inputs:
    - state
    - upgrade-database-flag
    state: stopped
```

Comment (wallyworld) [reply cmd/jujud/agent/machine/manifolds.go:414]: maybe add comments to these transitive ones to avoid confusion

Comment (manadart) [reply cmd/jujud/agent/machine/manifolds.go:415]: Done.

**Review by jameinel — APPROVED** [review]

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**